### ISSUE TITLE
services.mysqlBackup: Backup job might break due to garbage collection

### ISSUE THREAD

**Issue by fadenb** [NixOS/nixpkgs#29031]
## Issue description
Surprise email from cron daemon:
```
subject: Cron <root@HOSTNAME> mysql /nix/store/vsj6yazxlzrhwhcdvkzp2b1wakz96gpi-mariadb-10.1.25/bin/mysqldump DATABASENAME | /nix/store/c09pc52nm3ii38f6n7qx9k5g81d55w7l-gzip-1.8/bin/gzip -c > /var/backup/mysql/DATABASENAME.gz

execl: couldn't exec `/nix/store/c280fqkd65rrpp5mfarp6cc0v4ns6rlb-bash-4.4-p12/bin/bash'
execl: No such file or directory
```

The cron job created by the backup service is broken. Several referenced paths do no longer exist:
```
[root@HOSTNAME:~]# crontab -l
# DO NOT EDIT THIS FILE - edit the master and reinstall.
# (/etc/crontab installed on Tue Aug 22 18:01:15 2017)
# (Cron version V5.0 -- $Id: crontab.c,v 1.12 2004/01/23 18:56:42 vixie Exp $)
SHELL=/nix/store/c280fqkd65rrpp5mfarp6cc0v4ns6rlb-bash-4.4-p12/bin/bash
PATH=/nix/store/qladnhrhvlcxmzv2j2g7wybfj1mmzi7w-system-path/bin:/nix/store/qladnhrhvlcxmzv2j2g7wybfj1mmzi7w-system-path/sbin

NIX_CONF_DIR=/etc/nix
15 01 * * * mysql /nix/store/vsj6yazxlzrhwhcdvkzp2b1wakz96gpi-mariadb-10.1.25/bin/mysqldump  DATABASENAME | /nix/store/c09pc52nm3ii38f6n7qx9k5g81d55w7l-gzip-1.8/bin/gzip -c > /var/backup/mysql/DATABASENAME.gz



[root@HOSTNAME:~]# ls -al /nix/store/vsj6yazxlzrhwhcdvkzp2b1wakz96gpi-mariadb-10.1.25/bin/mysqldump
ls: cannot access '/nix/store/vsj6yazxlzrhwhcdvkzp2b1wakz96gpi-mariadb-10.1.25/bin/mysqldump': No such file or directory

[root@HOSTNAME:~]# ls -al /nix/store/c280fqkd65rrpp5mfarp6cc0v4ns6rlb-bash-4.4-p12/bin/bash
ls: cannot access '/nix/store/c280fqkd65rrpp5mfarp6cc0v4ns6rlb-bash-4.4-p12/bin/bash': No such file or directory

[root@HOSTNAME:~]# ls -al /nix/store/qladnhrhvlcxmzv2j2g7wybfj1mmzi7w-system-path/bin
ls: cannot access '/nix/store/qladnhrhvlcxmzv2j2g7wybfj1mmzi7w-system-path/bin': No such file or directory

```

Given enough time I will eventually reimplement the service using systemd-timers (as it was already discussed in #24728) but I might not get this done in time for the next release.

edit: It seems to me that services.cron.systemCronJobs it severly broken. Changes are sometimes not detected and cron runs for quite some time with an old version of the crontab (or no crontab at all!)

### Steps to reproduce
* Configure services.mysqlBackup
* ? (not sure what I did in between)
* nix-collect-garbage -d
* Get surprise email from cron telling you the backup failed

## Technical details

* System: 18.03.git.977073d (Impala)
* Nix version: nix-env (Nix) 1.11.14
* Nixpkgs version: "17.03.1247.05126bc850"

**Comment by fadenb** [comment]
Did some very quick changes as I needed to fix the backup for one of my hosts, :construction: work in progress can be found at https://github.com/fadenb/nixpkgs/commits/mysqlBackup_systemd_timers

Open issues:
* Broken authentication (#24728) (@florianjacob did you by any chance already start to implement the socket auth?)

**Comment by florianjacob** [comment]
@fadenb not yet. ☹️ As shown in #24728, the MySQL commands needed to set up socket auth are trivial, and I'm running it the way I showed there with a permission-reduced user successfully for a month now - with manually setting this up, though. The problem is doing this automatically.

There are two ways to get this set up automatically: Either through doing some MySQL bash script hackery [like the wordpress module is doing](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/apache-httpd/wordpress.nix#L266), as well as some other modules, or making that unnecessary by providing proper NixOS options infrastructure to set up MySQL users, and therefore fixing https://github.com/NixOS/nixpkgs/pull/6963 . Sadly, the current options of the MySQL module are useless for the backup script case as well.

The latter is the route I'd want to go, ditching the problem of password handling by making socket auth the default for all users created through some `mysql.users` option. But I'm unsure how to solve this problem brought up by copumpkin:

> Is there some way to make it clear to users that it won't delete databases? I could see it being both > scary (you'll delete my data) and counterintuitive (why doesn't my database list match my configuration.nix?)

What do you think?

**Comment by florianjacob** [comment]
PR for declarative Unix socket authentication at #29387

**Comment by rvl** [comment]
Hi @fadenb, I'm a little confused because there is `crontab -l` in your example.
However this command would show the file `/var/cron/tabs/root` -- the root user crontab.

NixOS edits the file `/etc/crontab`.

I am testing with the following steps:
1. In a separate terminal, watch the output of `journalctl -u cron.service -f`
2. Edit `services.mysqlBackup.databases` setting in `configuration.nix`.
3. `nixos-rebuild switch`
4. Note the log output -- the text `RELOAD (/etc/crontab)` should appear.

Could it be possible that the root *user* crontab (this is separate from `/etc/crontab`) has accidentally gotten the above contents? I'm not sure how this might happen, but could you please clarify for me.

**Comment by fadenb** [comment]
I actually did not even know cron was in use on that system before the mail with the error :p . So if that ended up in the wrong file it must have been due to something nix related.

I can no longer reproduce the issue as I already migrated that system to a systemd-timer based backup job.

**Comment by florianjacob** [comment]
@fadenb what's your plan on posing a PR for the systemd-timer change? Is there something you still want to change? I think it turned otu well, my only remark would be that I'd personally have set `AccuracySec = "5m";` to something larger like 1h, to follow the `systemd.timer` man page “To optimize power consumption, make sure to set this value as high as possible and as low as necessary.” - but I guess how much Accuracy is necessary differs between use cases. 🙃

Together with https://github.com/NixOS/nixpkgs/pull/29703 we could get this fixed finally. 👍

**Comment by rvl** [comment]
@fadenb I have merged your systemd-timer config into #29703.

This still leaves the crontab issue, which is of general interest for the release.

If you add
```
services.cron.systemCronJobs = [ "* * * * * root true" ];
```
to your config, is `/etc/crontab` updated correctly, and what is the result of running `crontab -l` as root?

**Comment by globin** [comment]
Fixed in #29703
